ooni/agent/scheduler.py

```diff
@@ -0,0 +1,409 @@
+import os
+import errno
+
+from hashlib import md5
+from datetime import datetime
+
+from twisted.application import service
+from twisted.internet import defer
+from twisted.internet.task import LoopingCall
+from twisted.python.filepath import FilePath
+
+from ooni.scripts import oonireport
+from ooni import resources
+from ooni.utils import log, SHORT_DATE
+from ooni.utils.files import human_size_to_bytes, directory_usage
+from ooni.deck.store import input_store, deck_store, DEFAULT_DECKS
+from ooni.settings import config
+from ooni.contrib import croniter
+from ooni.contrib.dateutil.tz import tz
+
+from ooni.geoip import probe_ip
+from ooni.measurements import list_measurements
+
+class FileSystemlockAndMutex(object):
+    """
+    This is a lock that is both a mutex lock and also on filesystem.
+    When you acquire it, it will first block on the mutex lock and then
+    once that is released it will attempt to acquire the lock on the
+    filesystem.
+
+    It's a way to support concurrent usage of the DeferredFilesystemLock
+    without races.
+    """
+    def __init__(self, file_path):
+        self._fs_lock = defer.DeferredFilesystemLock(file_path)
+        self._mutex = defer.DeferredLock()
+
+    @defer.inlineCallbacks
+    def acquire(self):
+        yield self._mutex.acquire()
+        yield self._fs_lock.deferUntilLocked()
+
+    def release(self):
+        self._fs_lock.unlock()
+        self._mutex.release()
+
+    @property
+    def locked(self):
+        return self._mutex.locked or self._fs_lock.locked
+
+# We use this date to indicate that the scheduled task has never run.
+# Easter egg, try to see what is special about this date :)?
+CANARY_DATE = datetime(1957, 8, 4, tzinfo=tz.tzutc())
+
+
+class DidNotRun(Exception):
+    pass
+
+
+class ScheduledTask(object):
+    _time_format = "%Y-%m-%dT%H:%M:%SZ"
+    schedule = None
+    identifier = None
+
+    def __init__(self, schedule=None, identifier=None,
+                 scheduler_directory=config.scheduler_directory):
+        if schedule is not None:
+            self.schedule = schedule
+        if identifier is not None:
+            self.identifier = identifier
```

Comment:
Let me see if I can give you a couple of examples as to why this is like this and perhaps you have some idea on how the code can be improved.

There are mainly 2 types of uses for the constructor of this class:

1. We use the default class attributes for schedule and identifier (this is the most common case), for example with the following:

* `UploadReports` (https://github.com/TheTorProject/ooni-probe/blob/v2.0.0-alpha/ooni/agent/scheduler.py#L414 and https://github.com/TheTorProject/ooni-probe/blob/v2.0.0-alpha/ooni/agent/scheduler.py#L149)
* `RefreshDeckList` (https://github.com/TheTorProject/ooni-probe/blob/v2.0.0-alpha/ooni/agent/scheduler.py#L263 and https://github.com/TheTorProject/ooni-probe/blob/v2.0.0-alpha/ooni/agent/scheduler.py#L417)

2. We need to change the identifier and the schedule depending on some value we read.
This is actually only the case for the RunDecks where the schedule and identifier is determined based on the deck name and the schedule written inside of the deck itself: https://github.com/TheTorProject/ooni-probe/blob/v2.0.0-alpha/ooni/agent/scheduler.py#L384
For this reason we need to support specialising at construction time these class attributes.